wallet/src/de/schildbach/wallet/service/WalletTransactionMetadataProvider.kt

```diff
@@ -99,23 +113,222 @@ class WalletTransactionMetadataProvider @Inject constructor(
     }
 
     override suspend fun setTransactionMemo(txId: Sha256Hash, memo: String) {
-        TODO("Not yet implemented")
+        updateAndInsertIfNotExist(txId) {
+            transactionMetadataDao.updateMemo(txId, memo)
+        }
+    }
+
+    override suspend fun setTransactionService(txId: Sha256Hash, service: String) {
+        updateAndInsertIfNotExist(txId) {
+            transactionMetadataDao.updateService(txId, service)
+        }
+    }
+
+    override fun syncTransactionBlocking(tx: Transaction) {
+        runBlocking {
+            syncTransaction(tx)
+        }
+    }
+
+    override suspend fun syncTransaction(tx: Transaction) {
+        log.info("sync transaction metadata: ${tx.txId}")
+        val metadata = transactionMetadataDao.load(tx.txId)
```

Comment:
Now there are two of these functions.  The first can be called from BlockchainServiceImpl in a dedicated handler.

The second is called by the first, but may at a later point in development be called by a worker that will gather historical transaction metadata (exchange rates).

Not sure if Blocking at the end of the name is the best nomenclature, perhaps it could be "Sync" though there is already sync in the name.

Reply:
yeah Blocking is fine